### PR TITLE
maint: add custom sampler warning

### PR DESCRIPTION
**Pull request by pkanal**
## Which problem is this PR solving?
Adds a warning if a custom sampler is provided. There are some pitfalls with providing a custom sampler if the `sampleRate` is non-deterministic.

## Short description of the changes
- Added a warning if `sampler` option is provided by the user

## How to verify that this has the expected result
- Unit tests pass 
- Test it out by adding a custom sampler to the `sampler` property in a test app.

### CODE REVIEW

File: src/honeycomb-options.ts [hunk 1]
```diff
@@ -22,7 +22,8 @@ export const MISSING_SERVICE_NAME_ERROR =
   'WARN: Missing service name. Specify either OTEL_SERVICE_NAME environment variable or serviceName in the options parameter.  If left unset, this will show up in Honeycomb as unknown_service:node';
 export const SKIPPING_OPTIONS_VALIDATION_MSG =
   'DEBUG: Skipping options validation. To re-enable, set skipOptionsValidation option or HONEYCOMB_SKIP_OPTIONS_VALIDATION to false.';
-
+export const SAMPLER_OVERRIDE_WARNING =
+  'WARN: Default deterministic sampler has been overridden. Make sure the sampleRate option is set and matches the sampleRate of  the custom sampler. Non-deterministic sampleRate could lead to missing spans in Honeycomb.';
```
Comment (JamieDanielson) [inline src/honeycomb-options.ts:26]:
```suggestion
  'WARN: Default deterministic sampler has been overridden. Honeycomb requires a resource attribute called SampleRate to properly show weighted values. Non-deterministic sampleRate could lead to missing spans in Honeycomb. See our docs for more details.';
```

**Review by JamieDanielson — COMMENTED** [review]
I suggested a change to the warning, which maybe could be tuned further.

It's important that the Resource Attribute is set for Honeycomb to properly weight it. And our sampleRate option in the config is only used for our Deterministic Sampler, so updating that doesn't work properly if overwriting the sampler.

Additionally, if setting up a new TraceIdRatioBasedSampler, the sample rate must be an integer N whereas the argument to the sampler is a ratio of 1/N ([docs blurb](https://docs.honeycomb.io/getting-data-in/opentelemetry/node-distro/#sampling-without-the-honeycomb-sdk)). So I was trying to think of a way to be sure the warning adds additional nuance without being too wordy.

Basically... there be dragons. Maybe that should be the warning 😅 

Here is some testing with 3 different samplers. Without updating the Resource, the Sample Rate (in Usage Mode) for non Honeycomb Deterministic Sampler is always 1, even if it's set on the SampleRate option in the config.

```ts
const config: HoneycombOptions = {
  apiKey: process.env.HONEYCOMB_API_KEY || '',
  serviceName: process.env.OTEL_SERVICE_NAME || 'hello-node-express-ts',
  debug: true,
  // sampler: new AlwaysOnSampler(),
  // sampler: new TraceIdRatioBasedSampler(0.5),
  sampleRate: 50,
  resource: new Resource({
    'mah-sampler': 'DeterministicSampler(honeycomb)',
    // sampleRate: 50,
  }),
};
```

mah-sampler-results

**Comment by pkanal** [comment]
> I suggested a change to the warning, which maybe could be tuned further.
> 
> It's important that the Resource Attribute is set for Honeycomb to properly weight it. And our sampleRate option in the config is only used for our Deterministic Sampler, so updating that doesn't work properly if overwriting the sampler.
> 
> Additionally, if setting up a new TraceIdRatioBasedSampler, the sample rate must be an integer N whereas the argument to the sampler is a ratio of 1/N ([docs blurb](https://docs.honeycomb.io/getting-data-in/opentelemetry/node-distro/#sampling-without-the-honeycomb-sdk)). So I was trying to think of a way to be sure the warning adds additional nuance without being too wordy.
> 
> Basically... there be dragons. Maybe that should be the warning 😅
> 
> Here is some testing with 3 different samplers. Without updating the Resource, the Sample Rate (in Usage Mode) for non Honeycomb Deterministic Sampler is always 1, even if it's set on the SampleRate option in the config.
> 
> ```ts
> const config: HoneycombOptions = {
>   apiKey: process.env.HONEYCOMB_API_KEY || '',
>   serviceName: process.env.OTEL_SERVICE_NAME || 'hello-node-express-ts',
>   debug: true,
>   // sampler: new AlwaysOnSampler(),
>   // sampler: new TraceIdRatioBasedSampler(0.5),
>   sampleRate: 50,
>   resource: new Resource({
>     'mah-sampler': 'DeterministicSampler(honeycomb)',
>     // sampleRate: 50,
>   }),
> };
> ```
> 
> mah-sampler-results

Thanks so much for the investigation! I think the docs could be updated as well in tandem with our change to allow overrides to the default sampler. I'll update the docs, change the wording here and then link to the docs specifically.

**Comment by pkanal** [comment]
@JamieDanielson On second thought, I'm going to link to the current docs for now because we'd be in a chicken / egg situation for the links if we have to wait for the release but this PR also has to be in the release. I think its easiest to link to the current docs and change it in the future.

**Review by JamieDanielson — APPROVED** [review]
